packages/boot/package.json

```diff
@@ -52,5 +52,10 @@
     "type": "git",
     "url": "https://github.com/strongloop/loopback-next.git",
     "directory": "packages/boot"
+  },
+  "peerDependencies": {
+    "@loopback/core": "^2.9.1",
+    "@loopback/repository": "^2.9.0",
+    "@loopback/service-proxy": "^2.3.4"
```

Comment:
This is potentially tricky.

Before my change, it was possible to use `@loopback/boot` in applications that don't have `@loopback/service-proxy` in their dependencies. `@loopback/boot` would install it's own copy that would be pretty much ignored at runtime.

With the change in place, apps must install both `@loopback/repository` and `@loopback/service-proxy` to be able to use `@loopback/boot`. I wish there was something like `optionalPeerDependencies`!

I see this as a sign that `@loopback/boot` is providing too many features and should be split into smaller packages.
1. It defines contract for Booter implementations and few base classes to help with implementation.
2. It implements the bootstrapping algorithm (`BootMixin`).
3. It provides booters for common artifacts (controllers, repositories, services, etc.).

I think the following structure would be better:
1. A new module to define Booter contract and possibly base classes. Prior art: https://github.com/strongloop/loopback-next/tree/master/packages/model-api-builder
2. `@loopback/boot` should provide `BootMixin` and artifact mixins that can be loaded using `@loopback/core` only.
3. Artifact booters requiring optional extensions (repository booter, service booter, etc.) should be moved either to those extensions (`@loopback/repository`, `@loopback/service-proxy`) or to a new package. A question to consider: where to put datasource booter that's shared by repositories and service proxies?

Such change is out of scope of this pull request though. The question is how to address the issue above inside this pull request:

1. Are we ok to require apps to install both `repository` and `service-proxy` if they want to use `@loopback/boot`? By default, `lb4 app` includes both, so it may be ok. But then it feels a kind of lame to force users to install packages they don't need.
2. Should I revert the repository & service-proxy part of this pull request and keep those two packages as regular `dependencies`?
3. We can also investigate how to rework booters to not require direct dependency on repository & service-proxy packages. They mostly call application methods mixed in by the user (e.g. `app.repository`) and import only type definitions for internal consumption from repository/service-proxy, so maybe it's ok to declare them in `devDependencies` only? The only exception I found is `ModelMetadataHelper.getModelMetadata`, it's called by `ModelBooter` to check if a class is decorated with `@model()`.

Reply:
That's my intention for https://github.com/strongloop/loopback-next/pull/5618.

Reply:
To move this pull request forward, I am going to pick the option 2:

>  revert the repository & service-proxy part of this pull request and keep those two packages as regular dependencies

It's not ideal, but it will give us more time to consider different options and implement a proper long-term fix.